Tweak speaker note about move closures and FnOnce

mauvealerts opened the pull request:
Fixes #651. I was going to just delete the note, but the disconnect is interesting. The current draft highlights it.

Example of a `move` closure implementing `Fn`
```rust
let x = String::new();
let x_len = move || x.len();
x_len();
x_len(); // OK
```

Example of a non-`move` closure which implements only `FnOnce`
```rust
let x = String::new();
let just_drop = || drop(x); // implicitly captures by value
just_drop();
just_drop(); // error
x.len(); // error
```

mgeisler reviewed (commented):
Thanks for bringing this up! This slide is one I don't like very much when teaching the class — the topic is complicated and needs much more time than we devote to it. So I would love to see it improved somehow (not sure how).

> `move` closures only implement `FnOnce`.

Now that I see this again, I also think it should go. I _think_ I was thinking of a situation where you `move` a variable into the closure _and_ you call a method/function which takes ownership. Like your `drop` example. However, my misunderstand was that `move` is not necessary then: Rust will automatically capture the variable with a `move` in that case.

Instead of the weaker (but correct!) statement that both `move` and non-`move` closures can be `FnOnce`, I think we should show a concrete example where `move` is needed.

What normally bites me is that I want to return a closure which only needs to borrow the captured variable — but because I return the closure, I need to `move` the variable explicitly. I would appreciate getting that example into the notes to make it clearer for everybody how to trigger the different cases.

djmitche commented:
I agree that we give this topic too little attention. It's a good place to synthesize students' understanding of move vs. copy, lifetimes, borrows, and auto-Deref.

Would it make sense to write an exercise for this?  Something with, at its core, "write Fn, FnOnce, and FnMut closures with and without `move`".  Or maybe the exercise gives six examples of callbacks and asks students to describe which Fn* types they satisfy and why.

Maybe I've just been reading too many math textbooks lately :)

mauvealerts commented:
Inspired by the feedback, I reworked things quite a bit:
* Expanded example to include closures of each type
* Reordered speaker notes to like up with examples
* Added notes about
  *  Inferred traits
  * `move`

I recall presenting this slide and, despite the complexity of the topic, not having much to "play with." This might be too much, though.

mauvealerts commented:
Oh, and I'm fine with adding an exercise around closures. I just haven't thought much about it yet.

mauvealerts commented:
If we want to stick to one slide and trim speaker notes a bit, I think I'd cut either/both:
* Note about `Copy` and `Clone`. It's a small detail in general. I mostly included to explain why `add_3` can be passed twice
* Clarification about subtyping ("I.e. you can use an..."). It's a lot of text and presumably folks are familiar with subtyping

If we want to split this into multiple slides, I suggest adding a short "Closures" section with a split like:
* `fn()` pointers vs `Fn` - emphasis on compatability
* Captures and `move`
* `Fn`, `FnMut`, `FnOnce`

mgeisler reviewed (approved):
Awesome, this is a much more well-rounded explanation of this :smile:

mgeisler commented:
> If we want to stick to one slide and trim speaker notes a bit

As a general rule, I consider the speaker notes to have "unlimited" space since they're not shown during the class. Even if they become long, people who read them outside of a class can pop them open in a separate window and thus have the notes next to the slide itself.

Also, as you can tell from this, the notes are called _Speaker Notes_, but I hope they are useful for people who read the course to learn Rust too. That dual-purpose is probably making us write strange notes at times and it's something I'm not 100% sure how to solve. Now that I write that, perhaps I should start by changing the header to _Notes_ instead... :thinking:

mgeisler commented:
> If we want to split this into multiple slides, I suggest adding a short "Closures" section with a split like:
> 
> * `fn()` pointers vs `Fn` - emphasis on compatability
> * Captures and `move`
> * `Fn`, `FnMut`, `FnOnce`

That sounds like a very good idea, actually! I am always in trouble when I get to the current slide: I am running out of time and don't want to get into this big topic. But... of course I do and people have lots of questions. So I end up trying to explain the topic using on-the-fly code and by showing people the types from the stdlib documentation.

So that's probably the worst of both worlds :sweat_smile: Having dedicated slides would be great since I can then skip them if we're out of time. @randomPoison and @fw-immunant, you've both been teaching the class repeatedly. Feel free to jump in with your feedback here!
